### PR TITLE
[GR11] Shuttle Excel connector code for BRI-1

### PR DESCRIPTION
**Pull request by biscuitdey**
# Description

Excel Addin that can be used with BRI-1 
This PR includes changes made for security improvements.

1. Email Validation - Regex match for valid email format
2. Password Validation - Regex match for valid password format
3. Secure storage -Storing and retrieving encrypted user information in IndexedDB
4. Secure IndexedDB - Implemented encryption and encoding to be used with IndexedDB using Web Cryptography API.
5. Form input validation - Regex match for valid inputs
6. HTML encoding - HTML encoding for input received from client (excel sheet) and server that will be displayed using .innerHTML() to prevent XSS attacks
7. JSON validation - Error catching and validation for JSON.parse()
 
## Related Issue

[[GR11] Security Improvements for Excel Addin](https://github.com/eea-oasis/baseline-grants/issues/33)

## Motivation and Context

This PR is related to the grant given to me for Excel-addin. The Microsoft Office library being used for development has multiple security issues which makes the Addin unfit for production. The library was selected as it is the official library for creating Addins and provides on-going support.
Known Issues:

- The Addin stores the user credentials within the Excel document causing these to be passed along whenever the user shares their Excel document with a counterparty. Currently, the Addin does not permit data to be stored on the local file system (needs to be developed).
- Addin allows anyone with access to the communication channels to tamper with the data.

## How Has This Been Tested

The Addin has been tested by locally running the Provide stack and interacting with it using Excel spreadsheets.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: examples/bri-1/lib/excel-connector/src/taskpane/excel-worker.ts [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+      onError(error.message);
+    } else {
+      onError(error);
+    }
+  }
+}
+
+export const excelWorker = new ExcelWorker();
+
+// function test() {
```
[the remark this answers is not in the document]
Comment (biscuitdey) [reply examples/bri-1/lib/excel-connector/src/taskpane/excel-worker.ts:86]: Test is obsolete. Will remove it.

File: examples/bri-1/lib/excel-connector/src/baseline/index.ts [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+      //Activate listeners on table or sheet
+      await this.activateListeners();
+
+      //Set Provide client for sending messages
+      this._identClient = identClient;
+
+      /*var data = {};
+      data["Col 1"] = "West";
+
+      const test = {
```
[the remark this answers is not in the document]
Comment (biscuitdey) [reply examples/bri-1/lib/excel-connector/src/baseline/index.ts:62]: Test is obsolete. Will remove it.

File: examples/bri-1/lib/excel-connector/cert.csr [hunk 3]
```diff
@@ -0,0 +1,48 @@
+-----BEGIN ENCRYPTED PRIVATE KEY-----
```
[the remark this answers is not in the document]
Comment (biscuitdey) [reply examples/bri-1/lib/excel-connector/cert.csr:1]: These key files are not required. Will remove it.

**Review by Therecanbeonlyone1969 — COMMENTED** [review]
do we still need the debugger? Also, only chrome and edge? No Brave support?

**Review by Therecanbeonlyone1969 — COMMENTED** [review]
@humbitious ... we now have an MIT license inside a CC0 license. I think we need a warning just because you have a CC0 license does not mean that you will not be sued by the license holder of modules used in the CC0 code. We should ask Chet how OASIS handles these type of situations.

**Review by Therecanbeonlyone1969 — COMMENTED** [review]
why do we need to print out the mapping?

File: examples/bri-1/lib/excel-connector/src/baseline/api.ts [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+      type: "mapping_type",
+      workgroup_id: workgroupId,
+      models: models,
+    };
+
+    console.log(mapping);
+
+    //TO SECURE --> JSON encoding
+    await identClient.createWorkgroupMapping(mapping);
+    //TODO: GET Mapping ID and do await store.setTableName(mappingTable, this.tableName) with mapping ID as mappingTable;
```
Comment (Therecanbeonlyone1969) [inline examples/bri-1/lib/excel-connector/src/baseline/api.ts:197]: Is this To Do still relevant? If yes, what is lost to no store the table name?

**Review by Therecanbeonlyone1969 — APPROVED** [review]
Several observations that we should be taking into account going forward, and only partially specific to this PR:
- large PRs are very hard to review. Prefer smaller, well-scoped PRs, and they can be draft PRs, but at least they can be reviewed well in a reasonable amount of time
- I  do not see separate tests with test fixtures, that should normally be a no-go. Will make n exception here
- I also prefer that tests are included in the github CI/CD pipeline through actions ... makes things a lot easier to review as well when I know all the tests past, so I can look at the tests first and see if they cover everything
- I also would like to suggest we use a linter such as eslint in a CI/CD action
- There are a lot of modules included that are boilerplate for e.g. NATS, ident, Oauth2, cert management, login with user name and password. These types of modules should be extracted into npm packages that can be utilized dynamically rather than having to be copied into an example itself. We are basically duplicating code that does not have to be duplicated and is creating tight coupling with modules that should be standard and in npm packages -> worthwhile big grant to do that work
- Specific to this PR, there are a lot of security to-dos -> recommend highlighting that in the read-me -> opportunity for another grant here

@humbitious @ognjenkurtic @rfisch -> should discuss in more detail during core-devs call

**Comment by biscuitdey** [comment]
@Therecanbeonlyone1969  Thank you for your review. Regarding the points you have raised:

1. The debugger is not required. I will remove it.
2. The browser for the Addin opens up inside the Excel Sheet. As per [Microsoft documentation](https://docs.microsoft.com/en-us/office/dev/add-ins/concepts/browsers-used-by-office-web-add-ins), there is only support Safari, Chrome, IE and Edge.
3. The mapping needn't be printed out. I will remove it.
4. Sorry, It seems there are some comments with TODO and TO SECURE that are still there in the code. The tasks have all been completed. I just need to take out the comments.

**Review by ognjenkurtic — APPROVED** [review]

**Comment by ognjenkurtic** [comment]
I had a look. Its really hard for the to review properly as i was not involved in this topic previously and the PR is quite big. Left some minor comments but it would be great if someone who is more in the topic could verify the PR. This is another great example of why we must go with smaller tasks -> smaller PRs.

File: examples/bri-1/lib/excel-connector/.vscode/settings.json [hunk 5]
```diff
@@ -0,0 +1,8 @@
+{
+    "eslint.validate": [
+      "javascript",
+      "javascriptreact",
+      "typescript"
+    ]
+  }
+
```
Comment (skarred14) [inline examples/bri-1/lib/excel-connector/.vscode/settings.json:8]: extra line

File: examples/bri-1/lib/excel-connector/.vscode/tasks.json [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+      "type": "npm",
+      "script": "start:desktop -- --app word",
+      "presentation": {
+        "clear": true,
+        "panel": "dedicated",
+      },
+      "problemMatcher": []
+    },
+    {
+      // To debug your Add-in:
```
Comment (skarred14) [inline examples/bri-1/lib/excel-connector/.vscode/tasks.json:72]: Do we need these comments?

File: examples/bri-1/lib/excel-connector/webpack.config.js [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+        fs: "empty",
+        // NOTE: To avoid - Module not found: Error: Can't resolve 'net' in 'C:\Work\EnvisionBlockchain\shuttle-excel\Repo\ShuttleExcel\node_modules\ts-nats\lib'
+        net: "empty",
+        // NOTE: To avoid - Module not found: Error: Can't resolve 'tls' in 'C:\Work\EnvisionBlockchain\shuttle-excel\Repo\ShuttleExcel\node_modules\ts-nats\lib'
+        tls: "empty",
+    }
+  };
+
+  return config;
+};
```
Comment (skarred14) [inline examples/bri-1/lib/excel-connector/webpack.config.js:145]: new line

**Comment by skarred14** [comment]
Left minor cosmetic comments. Overall, agree with @Therecanbeonlyone1969 and @ognjenkurtic that this is a large PR. 
1. Ideally the methods in the `lib/excel-connector` can be abstracted into interfaces in core/api or core/persistence, which can then be implemented under bri-1. 
2. Further, it would be great to add test cases - unit and integration tests as this goes across multiple services.
3. Documentation update or a flow diagram to indicate the system integration view.
4. +1 on adding linters and updating CI/CD (Actions). This may be more of a general tidying up of the overall codebase
Can be discussed regd priority or need for above updates in this or following PR's in core-devs call. @humbitious

**Comment by kthomas** [comment]
Thanks to @biscuitdey for this incredible effort... and to @ognjenkurtic, @Therecanbeonlyone1969 and @skarred14 for the reviews...

There is still some cleanup to be done here and there but this project has really taken shape in terms of how "connectors" can be developed for baseline...

Subsequent PRs that are more manageable to review will follow to do any of that cleanup.

@biscuitdey deserves full credit for this one... 🚀

**Review by kthomas — APPROVED** [review]

**Pull request merged**